pkg/controller/master/nodedrain/nodedrain_controller.go

```diff
@@ -513,3 +524,45 @@ func (ndc *ControllerHandler) listVMILabelMaintainModeStrategy(node *corev1.Node
 	return virtualmachineinstance.ListByNode(node, labels.NewSelector().Add(*req),
 		ndc.virtualMachineInstanceCache)
 }
+
+// checkDedicatedCpuPlacementRequirements checks if there are any VMs with
+// CPU pinning enabled. If the number of nodes with enabled `CPU Manager`
+// feature is not >= 2, then the function will return a list of VMs that
+// are impacted.
+func (ndc *ControllerHandler) checkDedicatedCpuPlacementRequirements(node *corev1.Node, vmiList []*kubevirtv1.VirtualMachineInstance) ([]string, error) {
+	var impactedVMI []string
+
+	// Skip if the `CPU Manager` feature is disabled on this node. This
+	// indicates there is no running virtual machine with enabled CPU
+	// pinning.
+	value, ok := node.Labels[kubevirtv1.CPUManager]
+	if ok && value != "true" {
+		return nil, nil
+	}
+
+	// Get the virtual machines with enabled CPU pinning.
+	for _, vmi := range vmiList {
+		if vmi.Spec.Domain.CPU != nil && vmi.Spec.Domain.CPU.DedicatedCPUPlacement {
+			impactedVMI = append(impactedVMI, namespacedVMName(vmi))
+		}
+	}
+
+	if len(impactedVMI) == 0 {
+		return nil, nil
+	}
+
+	// Get number of nodes with enabled `CPU Manager` feature. Exit
+	// immediately if there are more than or less than 1 node with this
+	// feature enabled.
+	selector := labels.Set{kubevirtv1.CPUManager: "true"}.AsSelector()
+	nodeList, err := ndc.nodeCache.List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list nodes with labels %s: %v", selector.String(), err)
+	}
+
+	if len(nodeList) != 1 {
```

Comment:
It may still not be enough to ensure that the VM can be migrated to another node.

e.g. peer node does not have enough CPU cores to pin, or CPU architecture mismatching ...

For those dynamic scheduling results directed by by `kubevirt` & `kubelet`, Harvester is hard to get a decision immediatly.

A further enhancement will be: the node-drain controller watches vmim instances, and if some vmims fail due to scheduling failure, then stop those VMs accordingly.

Reply:
should we not filter `nodeList` for current node before running the check? 

A better check would be `nodeList >= 1` after applying the filter or `nodeList >= 2` without the filter before we return?